README.md

```diff
@@ -338,6 +339,31 @@ function stateReducer(state, changes) {
 }
 ```
 
+### keyDownHandlers
+
+> `{ ...handlers }` | optional
+
+NOTE: Normally if you want to add keydown handlers to the button or input, you can simply pass `onKeyDown` to `getInputProps` and `getButtonProps` (respectively). However, if you want to override built-in handling of specific keys, you can use this prop.
+
+Also if you want to override the built-in behavior of _all_ the keys, you can add `event.preventDefault()` in your own handler and that will signal to downshift to skip it's built-in event handler.
+
+This prop allows you to use your own keydown handlers which can be handy if you want custom behavior. If a key and handler you pass is already defined in downshift, the custom handler you pass will be used instead.
+
+For example, if you create an Enter handler, this will not change the default behavior of the other keyDown events: ArrowUp, ArrowDown, and Escape.
+
+```jsx
+const handlers = {
+  Enter(event) {
+    // Custom behavior for ENTER keydown
+  },
```

Comment:
> [The intention is to allow you to completely override the behavior for a specific key. If you want downshift to merge behavior, then you'll use an onKeyDown event handler yourself.](https://github.com/paypal/downshift/pull/346#discussion_r170705698)

Does this not have any unforeseen consequences though? I see that you've mentioned using `onKeyDown` within `getInputProps` as a solution for both both #330 and #336, so I'm wondering when it would be a good idea to completely override the handler for a specific key.

I imagine that the code run within the default handlers is important to how Downshift works. By completely overriding it, the user (perhaps unintentionally) is forgoing any and all functionality that Downshift provides. That may sound obvious, but if you look at what things like #330 and #336 are attempting, it is not to remove all of Downshift's functionality; #330 is looking to get around the `event.preventDefault()` that stops the form submission when Downshift is open, but it does not wish to forgo Downshift's functionality to select the highlighted item if there is one.

I guess at this point I'm curious when it benefits the developer to use `keyDownHandlers` over a solution like adding `onKeyDown` to `getInputProps`.

Reply:
So, to prevent downshift from running its internal behavior, you must call `event.preventDefault()`. However, as you mentioned, in #330, the behavior that we don't like is the `event.preventDefault()` call. In #330, we don't the `event.preventDefault()` to be called at all. So we can neither leave things as they are nor prevent the default behavior.

> I imagine that the code run within the default handlers is important to how Downshift works.

Yes, it is, but when you're trying to customize how downshift works internally, you're expected to understand what the current behavior is. Luckily what downshift is doing for these is fairly simple (and is using functions that you can already use yourself anyway).

> By completely overriding it, the user (perhaps unintentionally) is forgoing any and all functionality that Downshift provides.

People who don't like how downshift works already have an understanding of the functionality downshift provides (that's a natural pre-requisite to not liking it :wink:).

When using `keyDownHandlers` you _will_ quite possibly duplicate logic in your own use case, but I don't see that as a problem. I see the `keyDownHandlers` use cases as very limited in scope, but ultimately quite powerful at the same time.

The fact is that short of creating a special prop for this use case specifically, there's no other way to accomplish this. And I'd much rather solve this problem using a more generally useful prop than a one-off prop and adding a bunch of conditional logic in the code. This attitude is also what lead to the `stateReducer` prop which provides a great deal more value than solving only the original problem it was intended to solve.

> I guess at this point I'm curious when it benefits the developer to use keyDownHandlers over a solution like adding onKeyDown to getInputProps.

To be brief, you'd use `keyDownHandlers` when you want to prevent the default behavior of the built-in keyDown handlers for a specific key, but keep the existing behavior for all other keys. You would also use it if you want to override the existing behavior of the keyDown handlers without calling `event.preventDefault()`. The use cases for these is pretty narrow, but it's the only way that I like. :)

Reply:
Just to understand where you're coming from a bit better, you're saying the use-case for this `keyDownHandlers` prop is as follows:

> you'd use `keyDownHandlers` when you want to prevent the default behavior of the built-in keyDown handlers for a specific key, but keep the existing behavior for all other keys. You would also use it if you want to override the existing behavior of the keyDown handlers without calling `event.preventDefault()`.

To break that up, let me look at the cases one at a time.

> to prevent the default behavior of the built-in keyDown handlers for a specific key, but keep the existing behavior for all other keys.

To consider an alternative, this can also be done already with `onKeyDown`.

```jsx
<input 
  {...getInputProps({
    onKeyDown: e => {
      if (e.key === 'Enter') {
        e.preventDefault()
        // prevent custom behavior for ENTER;
        // keep existing behavior for all other keys
      }
    }
  })}
/>
```

> to override the existing behavior of the keyDown handlers without calling `event.preventDefault()`

I would say this makes a stronger point than the first, but I would posit the question _why would you want to do this?_ Up until this point, #330 and #336 were the poster children for why a feature like `keyDownHandlers` might be desirable, but as we established, #330 wouldn't actually benefit from this feature, and #336 was actually just a question on why the default behavior is the way it is for this library (which I too am confused by; why is the "default" for holding the shift key to jump 5 indices?).

Even if we assume there is some narrow use-case for wanting to override the default behavior for all key down handlers, this can still be accomplished with `onKeyDown` by checking the event key.

Based on this, it seems like **anything you can do with `keyDownHandlers` you can already do—without any changes to the lib—using `onKeyDown` in `getInputProps`**.

On top of the functionality already being here, `onKeyDown` is more idiomatic with React (people already know they can provide an input element an `onKeyDown` handler); it requires no changes to the current code in Downshift (thus avoiding more docs to maintain and more custom props for people to understand); and it won't completely override what Downshift does under the hood unless you explicitly tell it to with `event.preventDefault()`.

Reply:
> People who don't like how downshift works already have an understanding of the functionality downshift provides

Why are we assuming that people who want to do something custom don't like what Downshift does already? What if they merely want a say in what happens on a type of event, but still want Downshift to do its own thing with the event as well? #330 is a perfect example of this.

It's just that the more I look at this feature, the less I'm convinced by any problem it is solving for people that they didn't have the ability to solve using `onKeyDown`. In fact, it _adds_ a layer of complexity because the developer must decide if what they really want is to completely override existing functionality or just add custom behavior on top of what exists.

If I'm fighting a losing battle and this feature is desirable no matter what, that's fine. At that point I would just strongly urge the docs to make a point that this is probably not what you want to use if you simply want to do something "extra" with a key down event. The only time you'd want this feature is if you want to _completely override_ Downshift's functionality to handle key down events.

Reply:
I think you're still misunderstanding things a bit...

> To consider an alternative, this can also be done already with onKeyDown.

Your example works, however doing that calls `event.preventDefault()` which will prevent the enter key from submitting the form, which is why #330 was brought up in the first place.

> as we established, #330 wouldn't actually benefit from this feature

That is not correct. I had attempted to establish that #330 would actually benefit from this feature.

The workaround for #330 without this feature is to re-implement the entire `onKeyDown` prop and override it entirely:

```javascript
<input
  {...getInputProps()}
  onKeyDown={customKeyDownThatImplementsNintyPercentOfDownshiftLogic}
/>
```

As the name implies, that custom key down handler will have to reimplement a bunch of stuff built into downshift just so it doesn't have to use `event.preventDefault()`. Specifically, what it'd have to duplicate is [this code](https://github.com/paypal/downshift/blob/5278c2c3c437e4861e7180d8a98160caff8872f7/src/downshift.js#L487-L517):

https://github.com/paypal/downshift/blob/5278c2c3c437e4861e7180d8a98160caff8872f7/src/downshift.js#L487-L517

They'd also have to reimplement `moveHighlightedIndex` (which I think we should probably expose in `getStateAndHelpers` anyway, so maybe not).

All that said, I'm starting to feel you have some good points. If we expose `moveHighlightedIndex`, then it's actually not a ton of code to duplicate. So I think I'd be fine with either. I'm not the one who has the use case though so I'd like to hear from folks who actually have this use case.

I think that we can boil down the decision to this:

1. Add `keyDownHandlers` as another way to do the same thing that does it in a slightly different way (which adds complexity to both downshift and its API)
2. Keep things as they are and if someone doesn't like that they have to use `event.preventDefault()` to prevent the downshift behavior, then they simply override things as I have shown above and have to reimplement any of the logic they _do_ want for their implementation. This makes it less ergonomic for people going against the grain.

Whatever we go with, I still think we should expose `moveHighlightedIndex`, so I didn't factor that into the options.

I'm leaning to option 2 now personally...

Reply:
> Your example works, however doing that calls `event.preventDefault()` which will prevent the enter key from submitting the form, which is why #330 was brought up in the first place.

I'm not sure I understand what you're saying here. Even without `event.preventDefault()` here in `onKeyDown`, Downshift will stop the form submission on ENTER regardless. AFAICT, _that_ was why #330 was brought up in the first place.

My example was to show that you could use `onKeyDown` "to prevent the default behavior of the built-in keyDown handlers for a specific key, but keep the existing behavior for all other keys", which was one of the points you brought up for `keyDownHandlers` as a prop.

If you wanted to submit the form, you could do that here as well.

```jsx
<input
  {...getInputProps({
    onKeyDown: e => {
      if (e.key === 'Enter') {
        this.props.onSubmit(e)
      }
    }
  })}
/>
```

> The workaround for #330 without this feature is to re-implement the entire `onKeyDown` prop and override it entirely

That is not necessarily true. Looking at the above snippet, one is able to submit a form _and_ retain the event handling provided by Downshift; the event will propagate from `onKeyDown` first and then travel on to Downshift's internal `keyDownHandlers`.

The whole benefit of `onKeyDown` (that I think I'm failing to explain properly) is that it lets you intercept the event, do what you want based on the event, and then let it continue on to `keyDownHandlers` if you so choose (otherwise you can stop it in `onKeyDown` by calling `event.preventDefault()`.

Reply:
I'm not 100% certain, but I feel like calling the `onSubmit` callback and submitting the form are two different things entirely.

> Also, is there a reason you have defined onKeyDown outside of getInputProps in your example?

The point of my example was to demonstrate how you could override the behavior without calling `event.preventDefault()` which is what you have to do today to make the form's submit event fire when the Enter key is pressed on the input regardless of the open state. _That_ is the reason #330 was opened in the first place. _That_ is the motivation behind this PR.

Reply:
> I'm not 100% certain, but I feel like calling the onSubmit callback and submitting the form are two different things entirely.

I'm not certain either, and I would be more inclined to agree in the scope of vanilla JS, but in React, the only way to get a form to submit is to pass it a submission handler, no?

```jsx
```

Reply:
This might be a naive question, but shouldn't the event only be restricted from propagating if the menu is open and there is an item highlighted? Otherwise the focus is still on the input itself, and I would expect an event to propagate like normal:

```
Enter(event) {
    if (this.getState().isOpen
      && this.getState().highlightedIndex) {
      event.preventDefault()
      this.selectHighlightedItem({
        type: Downshift.stateChangeTypes.keyDownEnter,
      })
    }
},
```

Reply:
> I'm not 100% certain, but I feel like calling the onSubmit callback and submitting the form are two different things entirely.

Yeah, I'm not certain either, but unless I'm mistaken, in React the only way to get a form to submit is to pass a submission handler, either to form itself or to a submit button within the form.

If there is a difference between invoking a callback on form submission and just "submitting the form", I'd be curious to know what it is.

Reply:
> The point of my example was to demonstrate how you could override the behavior without calling `event.preventDefault()` which is what you have to do today to make the form's submit event fire when the Enter key is pressed on the input regardless of the open state.

Hmm, yeah this makes sense to me know. I hadn't considered if a user was only providing an `onSubmit` handler to the form element itself. I guess it just felt strange to me to do this and not also invoke the same `onSubmit` handler in the `onKeyDown` event for ENTER.

This is how we're currently doing it

```jsx
<form onSubmit={this.handleSubmit}> // invoked if the user clicks the "Search" button
  <Downshift
    render={({ getInputProps, highlightedIndex }) => (
      <input
        {...getInputProps({
          onKeyDown: e => {
            if (e.key === 'Enter' && highlightedIndex) {
              this.handleSubmit(e) // invoked on ENTER key down
            }
          }
        })}
      />
    )}
  />
  <button type='submit'>Search</button>
</form>
```

Reply:
So in our application we have two different types of Downshift autosuggest inputs:

1. Autosuggest values, and force the user to make a selection before doing any kind of submit action.
2. Autosuggest values, but allow the user to submit a "custom" value that isn't selected from the list.

We also have deeply nested components that have abstractions for handling aria compliance, error handling etc., so we have been relying on normal event propagation to handle for submission. The eventual `onSubmit()` handler gathers up all the values from this and potentially other inputs and submits everything.

The #1 style form is easily supported by downshift, but the #2 style is hindered by the `event.preventDefault()`.

As another option, would it make sense to support a prop like `allowCustomValues` that would enable the behavior I mentioned above? This would allow both of these styles of forms.

I could also be wildly oversimplifying things, I am not familiar with all use cases for Downshift :)

Reply:
@ericedem you can refer to the Type Ahead example https://codesandbox.io/s/82m2px40q9 or https://codesandbox.io/s/1k3m67koj for your second use case.

Reply:
Hi Kent, sorry for the late response! I've been caught up with some work so I'm writing this at 12am 😴 . Thanks to everyone for contributing to a constructive conversation. I really appreciate it. I think everyone made some solid points in this thread. I would like to add my two cents after having read these conversations. 

My teammate and I have implemented a "workaround" solution addressing #330 in our autocomplete. Our workaround was the one @indiesquidge posted. He's actually my teammate and we'll both be at Kent's March testing workshops 😸 we are **excited**! For the sake of reference for all readers, here is our "workaround":

```jsx
<form onSubmit={this.handleSubmit}> {/* invoked on "Search" button click */}
  <Downshift
    render={({ getInputProps, highlightedIndex }) => (
      <input
        {...getInputProps({
          onKeyDown: e => {
            if (e.key === 'Enter' && highlightedIndex === null) {
              this.handleSubmit(e) // invoked on ENTER key down
            }
          }
        })}
      />
    )}
  />
  <button>Search</button>
</form>
```

I am now leaning towards **not using the keyDownEvents prop to solve issue #330**. The following statement made me reconsider:

>⚠ *This is an advanced feature and you should avoid using it* ⚠
>This feature exposes the downshift instance in a way that could lead to your code breaking if we decide to change a method you are relying on.

I would prefer the "workaround" that I have. I would have to do some logical checks to get a `<form>` to call the `onSubmit` handler on `ENTER`, but at least it is guaranteed to be safe. As stated above, I should probably avoid relying an exposed instance property. The current workaround is less code and more safe (e.g. it won't break when someone decides to rename `selectHighlightedItem` in downshift). 

In my use case, I'm only trying to invoke a the form's `onSubmit` handler. Like Kent said, I think this is completely different than a form submission. 

If we want to support narrow use cases to *completely* override Downshift's keydown handlers, then we could continue with this pull request. Otherwise, my personal vote is for option 2. Kent, if you're leaning towards option 2, then I can close this pull request.

> 2) Keep things as they are and if someone doesn't like that they have to use event.preventDefault() to prevent the downshift behavior, then they simply override things as I have shown above and have to reimplement any of the logic they do want for their implementation. This makes it less ergonomic for people going against the grain.

Reply:
Not sure if anyone still wants opinions on this, but I like @hzhu's implementation. Per the original issue #330, the main problem is it should be possible to have the form submission handler be called when Enter event is fired:

"I think that a <form> submission handler should always be called on ENTER, even if the combo box list is open."

Notice that the issue DIDN'T say "I think you should be able to use the default functionality of Enter within a form"

I don't think @hzhu actually wanted to use the default functionality when the form is submitted. I think he just wanted to invoke onSubmit which would look something like this:

```jsx

submitHandler(event) {
   // Do something with form contents
}

const handlers = {
   Enter(event) {
    // NO event.preventDefault()
  },
}

const ui = 
   <form onSubmit={submitHandler}>
      <Downshift keyDownHandlers={handlers} />
      <input /> // non-downshift input
      <input /> // non-downshift input
   </form>
```
If that is the case, then you certainly could just handle the "submit" in the custom Enter handler like this:

```jsx
const handlers = {
   Enter(event) {
    event.preventDefault()
    // Do something with form contents
  },
}

const ui = 
   <form>
      <Downshift keyDownHandlers={handlers} />
      <input /> // non-downshift input
      <input /> // non-downshift input
   </form>
```
But then why do this in comparison to @indiesquidge solution above? In my opinion, the answer to this question is because then you can grab the form data from the event if that is what you want (though from what I have read this is not recommended with react). You wouldn't have to have onChange handlers to grab the state of the non-downshift inputs nor would you need to use refs. You could just grab the values in your Enter event handler.

Reply:
@donysukardi Sorry, that doesn't address the issue, since not do an [implicit submission](https://www.w3.org/TR/html52/sec-forms.html#implicit-submission) on enter is the core problem here.

I think the workaround passing a submit handler into a an input works ok for simple use cases, but the nice thing about using the standard eventing structure is when you have multiple abstraction layers.

If I had a form using input components with two levels of abstraction it would need to look like this:

```
<form onSubmit={submitHandler}>
  <CustomInput onSubmit={submitHandler}/>
</form>

function CustomInput ({ submitHandler }) {
  // do some other things
  return <AutocompleteInput onSubmit={submitHandler} />
}

function AutocompleteInput ({ submitHandler }) {
  return (
    <Downshift>
      {({ getInputProps, highlightedIndex }) => (
      <input
        {...getInputProps({
          onKeyDown: e => {
            if (e.key === 'Enter' && highlightedIndex === null) {
              submitHandler(e) // invoked on ENTER key down
            }
          }
        })}
    </Downshift>
  )}
}
```

This, has the result of sprinkling knowledge of the form submission through several places. With events, the input doesn't need to know anything about what it is used for.

Reply:
@ericedem is correct. The implicit submission is an important thing we want to allow.

What I'm starting to think is that we need a downshift 2.0.0 that changes the way you prevent the built-in behavior of downshift. It should be a fairly trivial upgrade for people. I'm going to file a new issue for this so we can discuss it, but I think that would resolve the issue in #330.